lib/Dialects/LinalgExt/Transforms/TilingExternalModels.cpp

```diff
@@ -0,0 +1,113 @@
+//===- TilingExternalModels.cpp - External models for TilingInterface -----===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "Dialects/LinalgExt/Passes.h"
+
+#include "mlir/Dialect/Arithmetic/IR/Arithmetic.h"
+#include "mlir/Dialect/Linalg/IR/LinalgOps.h"
+#include "mlir/Dialect/Linalg/Utils/Utils.h"
+#include "mlir/Interfaces/TilingInterface.h"
+
+using namespace mlir;
+using namespace mlir::linalg;
+
+static Value getAsValue(OpBuilder &b, Location loc, OpFoldResult ofr) {
+  if (auto v = ofr.dyn_cast<Value>())
+    return v;
+  return b.create<arith::ConstantIndexOp>(
+      loc, ofr.get<Attribute>().cast<IntegerAttr>().getInt());
+}
+static SmallVector<Value> getAsValues(OpBuilder &b, Location loc,
+                                      ArrayRef<OpFoldResult> ofrs) {
+  SmallVector<Value> vals;
+  vals.reserve(ofrs.size());
+  for (auto ofr : ofrs)
+    vals.push_back(getAsValue(b, loc, ofr));
+  return vals;
+}
+
+namespace {
+
+/// External model implementation of TilingInterface for LinalgOps. This is
+/// templated on the actual Linalg named op for now since the registration of
+/// the external model requires the original operation.
+template <typename LinalgOpTy>
+struct LinalgOpTilingInterface
+    : public TilingInterface::ExternalModel<LinalgOpTilingInterface<LinalgOpTy>,
+                                            LinalgOpTy> {
+  SmallVector<Value> getDestinationOperands(Operation *op, OpBuilder &b) const {
+    LinalgOp linalgOp = cast<LinalgOp>(op);
+    return linalgOp.getOutputOperands();
+  }
+
+  SmallVector<StringRef> getLoopIteratorTypes(Operation *op) const {
+    LinalgOp linalgOp = cast<LinalgOp>(op);
+    SmallVector<StringRef> iteratorTypes;
+    iteratorTypes.reserve(linalgOp.iterator_types().size());
+    for (Attribute iteratorAttr : linalgOp.iterator_types()) {
+      iteratorTypes.push_back(iteratorAttr.cast<StringAttr>().getValue());
+    }
+    return iteratorTypes;
+  }
+
+  SmallVector<Range> getLoopBounds(Operation *op, OpBuilder &b) const {
+    LinalgOp linalgOp = cast<LinalgOp>(op);
+    return linalgOp.createLoopRanges(b, op->getLoc());
+  }
+
+  Operation *getTiledImplementation(Operation *op, OpBuilder &b,
+                                    ValueRange dest,
+                                    ArrayRef<OpFoldResult> offsets,
+                                    ArrayRef<OpFoldResult> sizes) const {
+    LinalgOp linalgOp = cast<LinalgOp>(op);
+    Location loc = op->getLoc();
+    AffineMap shapeSizesToLoopsMap = linalgOp.getShapesToLoopsMap();
+    auto allShapeSizes = linalgOp.createFlatListOfOperandDims(b, loc);
```

Comment:
Sorry, I didnt add any comments when I initially wrote this. Would it be possible to comment this a little bit?